docs/design/feature-attribute-semantics.md

```diff
@@ -0,0 +1,203 @@
+
+## Motivation
+
+Existing attributes like `RequiresUnreferencedCodeAttribute`, `RequiresDynamicCodeAttribute`, and `RequiresAssemblyFilesAttribute` have behavior close to what is described below. The behavior differs slightly between illink and NativeAot in the details, so this is an attempt to specify the semantics as clearly as possible, so that both tools can converge to match this.
+
+The ILLink Roslyn analyzer also produces warnings for these attributes, but doesn't have insight into the compilation strategy used for compiler-generated code. These rules are designed so that the warnings produced by a Roslyn analyzer are matched by the IL analysis, but IL analysis may include additional warnings (specifically for reflection access to compiler-generated code).
+
+There is also the possibility that we will create an attribute-based model which allows users to define their own feature attributes; see this draft for example: https://github.com/dotnet/designs/pull/261. The semantics outlined here could be extended to those attributes if we determine that they are appropriate there.
+
+We would also like to share as much code as possible for this logic between the linker, NativeAot, the corresponding analyzers, and possibly future analyzers.
+
+## Goals
+
+- Define the semantics of feature attributes
+- Define the access patterns which are allowed and disallowed by these semantics
+
+## Non-goals
+
+- Specify the warning codes or wording of the specific warnings for disallowed access
+- Define a model for defining new feature attributes
+- Define an attribute-based model for feature switches
+- Specify the relationship between feature switches and feature attributes
+- Define the interactions between `RequiresUnreferencedCodeAttribute` and `DynamicallyAccessedMembersAttribute`
+
+## RequiresFeatureAttribute
+
+`RequiresFeatureAttribute` may be used on methods, constructors, or classes only.
+
+The use of this attribute establishes a [_feature requirement_](#feature-requirement) for the attributed type or member, which restricts access to the attributed type or member (and in some cases to other related IL) in certain ways. It also establishes a [_feature available_](#feature-available-scope) scope (which includes the attributed member but may also include other related IL) wherein access to members with a _feature requirement_ is allowed.
+
+Access to members with a _feature requirement_ is always allowed from a _feature available_ scope, and never produces feature warnings. The restrictions created by _feature requirement_ only limit access from scopes outside of _feature available_, where certain access patterns produce warnings.
+
+## Feature available scope
```

Comment:
Content for this section looks good, but it's interesting that this is represented as bullets in the feature requirements section, but prose here.

For someone implementing the spec, you might want a more structured notation, i.e. the following constructs declared with a _feature requirement_ are also in a _feature available scope_: etc.